`str` on sumtypes doesn’t always work

When a type created with `deftype` wraps another type, the `str` function
that is automatically generated will use `%p` to print the type that was
wrapped—if there is no `str` function that is automatically defined on the
wrapped type. This does not always work if the type is a value type.

A motivating example:

```clojure
; registered types do not get an automatic `str`
(relative-include "mytype.h")
(register-type MyType)
(register to-my-type (Fn [Int] MyType))

(defn main []
  (println* &(Maybe.Just (to-my-type 1))))
```

To make this work, here’s an example version of `mytype.h`:

```c
typedef int MyType;

MyType to_MINUS_my_MINUS_type(int x) {
  return x;
}
```

On my version of Clang and Carp, this emits warnings like:

```
/Users/veitheller/.carp/out/main.c:2045:38: warning: format specifies type
      'void *' but the argument has type 'MyType' (aka 'int') [-Wformat]
        tempsize = snprintf(NULL, 0, "%p", p->Just.member0);
```

Also note that this does not get resolved by defining `str` or `prn` on the
reference type, as I would’ve expected. It is only resolved when `prn` is
defined for the type itself, breaking the assumption that those functions
take references.

I’m not quite sure what the best way to resolve this is. A quick fix for `Maybe`
would be to redefine `str` for it manually. That would only b a band-aid,
though. Ultimately, we only have a few principld options here (that I can see,
feel free to add more):

- Emit an error if the user tries to print a wrapped type that doesn’t have
  `str` defined. This would replace the default `printf` implementation
  entirely.
- Emit a value that makes clear that the value is unprintable.
- Move to `str` for referenc types—this should be done anyway, in my opinion,
  but it might not be ergonomic enough on its own.

As I see it, the first two are probably mutually exclusive, but the third one
should be done either way.

Cheers

--- a/src/StructUtils.hs
+++ b/src/StructUtils.hs
@@ -23,9 +23,8 @@ memberPrn typeEnv env (memberName, memberTy) =
                   ]
         Nothing ->
           if isExternalType typeEnv memberTy
-          then unlines [ "  tempsize = snprintf(NULL, 0, \"%p\", p->" ++ memberName ++ ");"
-                       , "  temp = malloc(tempsize);"
-                       , "  snprintf(temp, tempsize, \"%p\", p->" ++ memberName ++ ");"
+          then unlines [ "  temp = malloc(11);"
+                       , "  snprintf(temp, 11, \"<external>\");"
                        , "  snprintf(bufferPtr, size, \"%s \", temp);"
                        , "  bufferPtr += strlen(temp) + 1;"
                        , "  if(temp) { CARP_FREE(temp); temp = NULL; }"
@@ -44,7 +43,7 @@ memberPrnSize typeEnv env (memberName, memberTy) =
                  ]
        Nothing ->
          if isExternalType typeEnv memberTy
-         then unlines ["  size +=  snprintf(NULL, 0, \"%p \", p->" ++ memberName ++ ");"
+         then unlines ["  size +=  11;"
                       ,"  if(temp) { CARP_FREE(temp); temp = NULL; }"
                       ]
          else "  // Failed to find str function for " ++ memberName ++ " : " ++ show memberTy ++ "\n"

